plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCacheImpl.java

```diff
@@ -0,0 +1,208 @@
+                    (ns, offloader) -> {
+                        if (offloader != null && Objects.equals(offloader.getOffloadPolicies(), offloadPolicies)) {
+                            return offloader;
+                        }
+                        else {
+                            if (offloader != null) {
+                                offloader.close();
+                            }
+                            return initManagedLedgerOffloader(offloadPolicies, pulsarConnectorConfig);
+                        }
+                    });
+            managedLedgerConfig.setLedgerOffloader(ledgerOffloader);
+        }
+        return managedLedgerConfig;
+    }
+
+    private synchronized OrderedScheduler getOffloaderScheduler(OffloadPolicies offloadPolicies)
+    {
+        if (offloaderScheduler == null) {
+            offloaderScheduler = OrderedScheduler.newSchedulerBuilder()
+                    .numThreads(offloadPolicies.getManagedLedgerOffloadMaxThreads())
+                    .name("pulsar-offloader").build();
+        }
+        return offloaderScheduler;
+    }
+
+    private LedgerOffloader initManagedLedgerOffloader(OffloadPolicies offloadPolicies,
+                                                       PulsarConnectorConfig pulsarConnectorConfig)
+    {
+            if (!Strings.nullToEmpty(offloadPolicies.getManagedLedgerOffloadDriver()).trim().isEmpty()) {
+                checkNotNull(offloadPolicies.getOffloadersDirectory(),
+                        "Offloader driver is configured to be '%s' but no offloaders directory is configured.",
+                        offloadPolicies.getManagedLedgerOffloadDriver());
+                Offloaders offloaders = offloadersCache.getOrLoadOffloaders(offloadPolicies.getOffloadersDirectory(),
+                        pulsarConnectorConfig.getNarExtractionDirectory());
+                LedgerOffloaderFactory offloaderFactory = null;
+                try {
+                    offloaderFactory = offloaders.getOffloaderFactory(
+                            offloadPolicies.getManagedLedgerOffloadDriver());
+                } catch (IOException ioe) {
+                    log.error("Failed to fetch offloaderFactory: ", ioe);
+                    throw new RuntimeException(ioe.getMessage(), ioe.getCause());                }
+                try {
+                    return offloaderFactory.create(
+                            offloadPolicies,
+                            ImmutableMap.of(
+                                    LedgerOffloader.METADATA_SOFTWARE_VERSION_KEY.toLowerCase(), PulsarVersion.getVersion(),
+                                    LedgerOffloader.METADATA_SOFTWARE_GITSHA_KEY.toLowerCase(), PulsarVersion.getGitSha()),
+                            getOffloaderScheduler(offloadPolicies));
+                }
+                catch (IOException ioe) {
+                    log.error("Failed to create offloader: ", ioe);
```

Comment:
no point to log if you throw. Please fix other places like that too.

Reply:
updated.